// @olympusat/oly-client-sdk/lib/models/AppsModel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AppsModel = AppsModel;

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _graphql = require('../graphql');

var _graphql2 = _interopRequireDefault(_graphql);

var _fun = require('../utils/fun');

var _constants = require('../utils/constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function AppsModel(options) {
  var list = function list(args) {
    return _utils2.default.call(options.ENDPOINT_APPS, _graphql2.default.apps.list(), { key: "apps" });
  };

  var listActive = function listActive(user) {
    return new _bluebird2.default(function (resolve, reject) {
      var activeCache = window.Oly && window.Oly.meta.activeApps.length > 0 ? window.Oly.meta.activeApps : (0, _fun.storage_getItem)(_constants.USER_APPS_LS_KEY);

      // Return cached user apps
      if (activeCache && options.allowCaching) {
        resolve(activeCache);
        return true;
      }

      _utils2.default.call(options.ENDPOINT_APPS, _graphql2.default.apps.list(), { key: "apps" }).then(function (res) {
        var activeApps = res && res.length > 0 ? res.filter(function (app) {
          return app.isActive;
        }) : [];

        if (window.Oly && window.Oly.meta.activeApps) {
          window.Oly.meta.activeApps = activeApps;
        }

        (0, _fun.storage_setItem)(_constants.USER_APPS_LS_KEY, activeApps);

        resolve(activeApps);
      });
    });
  };

  var getById = function getById(_ref) {
    var entityUUID = _ref.entityUUID;
    return _utils2.default.call(options.ENDPOINT_APPS, _graphql2.default.apps.getById({ entityUUID: entityUUID }), { key: "app" });
  };

  var getBySlug = function getBySlug(_ref2) {
    var appSlug = _ref2.appSlug;
    return _utils2.default.call(options.ENDPOINT_APPS, _graphql2.default.apps.getBySlug({ appSlug: appSlug }), { key: "appBySlug" });
  };

  return {
    list: list,
    listActive: listActive,
    getById: getById,
    getBySlug: getBySlug
  };
}
